src/NumericalAlgorithms/LinearOperators/ApplyMatricesToVariables.cpp

```diff
@@ -0,0 +1,365 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "NumericalAlgorithms/LinearOperators/ApplyMatricesToVariables.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <vector>
+
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "NumericalAlgorithms/LinearOperators/Transpose.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/DereferenceWrapper.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+namespace {
+void multiply_in_first_dimension(const Matrix& matrix, const double* data,
+                                 const gsl::not_null<size_t*> data_size,
+                                 const gsl::not_null<double*> result) noexcept {
+  *data_size /= matrix.columns();
+  dgemm_<true>('N', 'N',
```

Comment:
This relies on `Matrix::data()` being stored column-major. Can we make this explicit, e.g. template `apply_matrix_to_variables` on the storage order of the `MatrixType`? We could make that a `bool` and align it with `blaze::columnMajor` and `blaze::rowMajor`.

Alternatively we could have a type `MeshOperator` (or similar) that is basically just `Matrix` renamed, but which explicitly documents its storage order and is only applicable to `DataVector`s and (through the functionality you introduce in this PR) to `Variables`. This would derive off `DenseMatrix<double, blaze::columnMajor>` as introduced in #696, and would be what the `Spectral` functions return.

With the `MatrixType` deriving off `blaze` we could even optimise this further for diagonal matrices that are marked as such by a [Blaze adaptor](https://bitbucket.org/blaze-lib/blaze/wiki/Adaptors).

Reply:
Basically nothing in this PR needs to change. We'd only rename `Matrix` to `MeshOperator` (suggestions for other names?) as part of #696.

Reply:
All the code that uses BLAS already assumes column-major, and I believe we've decided that is all we want to support for the time being.  (Although the diagonal optimization would be nice.)

What uses do you have for a matrix other than as a linear operator?